README.md
# lukemadera:autoform-googleplace

An add-on Meteor package for aldeed:autoform. Provides a single custom input type, "googleplace", which renders an input that is given Google Places Autocomplete functionality (but without any map; just a pure input box) and that returns an object with `fullAddress`, `lat`, `lng`, `street`, `city`, `state`, `zip`, `country`.
https://developers.google.com/maps/documentation/javascript/examples/places-autocomplete

This DOES work when wrapped in Phonegap / Cordova AND with iOS8 and 3rd party keyboards (which the default GoogleUI places autocomplete does NOT)!
More info:
- on iOS Cordova (only happens when wrapped in an app it seems), the autocomplete results can not be selected (a fastclick timing issue it seems)
  - http://stackoverflow.com/questions/9972080/cant-tap-on-item-in-google-autocomplete-list-on-mobile
- on iOS 3rd party keyboards (e.g. Swift), the autocomplete results do not show up at all as the (keyup) events do not fire
  - http://stackoverflow.com/questions/26080368/ios-8-3rd-party-keyboards-dont-register-javascript-jquery-keyup-keypress-keyd


## Demo

[Demo](http://lukemadera-packages.meteor.com/af-googleplace-basic)

[Source](https://github.com/lukemadera/meteor-packages/tree/master/autoform-googleplace/basic)


## Dependencies

- aldeed:autoform
- Google Maps Places script (put in your entry HTML file)
  - `<script type="text/javascript" src="https://maps.googleapis.com/maps/api/js?libraries=places"></script>`


## Installation

In a Meteor app directory:
```bash
meteor add lukemadera:autoform-googleplace
```


## Usage

Specify "googleplace" for the `type` attribute of any input and set the SimpleSchema to be an object:

```html
{{> afQuickField name="address" type="googleplace" opts=optsGoogleplace}}
```

In the schema, which will then work with a `quickForm` or `afQuickFields`:

```js
AddressSchema =new SimpleSchema({
  fullAddress: {
    type: String
  },
  lat: {
    type: Number,
    decimal: true
  },
  lng: {
    type: Number,
    decimal: true
  },
  geometry: {
    type: Object,
    blackbox: true
  },
  placeId: {
    type: String
  },
  street: {
    type: String,
    max: 100
  },
  city: {
    type: String,
    max: 50
  },
  state: {
    type: String,
    regEx: /^A[LKSZRAEP]|C[AOT]|D[EC]|F[LM]|G[AU]|HI|I[ADLN]|K[SY]|LA|M[ADEHINOPST]|N[CDEHJMVY]|O[HKR]|P[ARW]|RI|S[CD]|T[NX]|UT|V[AIT]|W[AIVY]$/
  },
  zip: {
    type: String,
    regEx: /^[0-9]{5}$/
  },
  country: {
    type: String
  }
});

PropertySchema =new SimpleSchema({
  address: {
    type: AddressSchema,
    optional: true
  }
});

Template.autoformGoogleplaceBasic.helpers({
  optsGoogleplace: function() {
    return {
      // type: 'googleUI',
      // stopTimeoutOnKeyup: false,
      // googleOptions: {
      //   componentRestrictions: { country:'us' }
      // }
    }
  }
});
```


### Options

See the [documentation at the top of the javascript file](https://github.com/lukemadera/meteor-autoform-googleplace/blob/master/lukemadera_autoform-googleplace.js) for options you can pass in to customize.



CHANGELOG.md
# 1.0.7 (2015-10-16)

## Features
- return `placeId` for Google `place_id` for looking up any additional information


# 1.0.6 (2015-08-02)

## Features
- add keyboard up, down, escape, enter support (closes #4)


# 1.0.5 (2015-08-01)

## Features
- add geoJSON support (closes #2)


# 1.0.4 (2015-05-22)

## Features
- refactor to support multiple instances (on same view / page - previously this did not work; could only use one at a time)


# 1.0.3 (2015-05-19)

## Features
- add `googleOptions` pass through option for customization and no longer default to US countries only


# 1.0.2 (2015-05-02)

## Bug Fixes
- change to default `stopTimeoutOnKeyup` option to false since some Android Cordova apps were not displaying results. Worse for performance but should fix this issue.


# 1.0.1 (2015-03-17)

## Features
- update for autoform v5.0

## Bug Fixes
- fix instance issue where dropdown would disappear or show old values sometimes after the first use


# 1.0.0 (2015-02-25)

## Features
- added iOS 3rd party keyboard support!
  - set the new boolean `stopTimeoutOnKeyup` option to false to ensure this works even when toggling back and forth between keyboards (by default, for performance, 3rd party keyboards work BUT as soon as a keyup is dectected, the timeout stops running so if a regular keyboard is used and THEN a 3rd party keyboard is used later, the 3rd party keyboard will no longer work anymore)
- option naming change: changed from `standard` to `googleUI` for using the standard GoogleUI `type` option
- styling (box-sizing on dropdown, colors)

package.js
Package.describe({
  name: 'lukemadera:autoform-googleplace',
  version: '1.0.7',
  // Brief, one-line summary of the package.
  summary: 'Google Places Autocomplete input (without map). Returns an object with formatted address components',
  // URL to the Git repository containing the source code for this package.
  git: 'https://github.com/lukemadera/meteor-autoform-googleplace',
  // By default, Meteor will default to using README.md for documentation.
  // To avoid submitting documentation, set this field to null.
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.0.3.1');
  api.use('templating@1.0.0');
  api.use('blaze@2.0.0');
  api.use('aldeed:autoform@4.0.0 || 5.0.0');
  api.use('reactive-var@1.0.5');
  api.addFiles([
    'lukemadera_autoform-googleplace.html',
    'lukemadera_autoform-googleplace.css',
    'lukemadera_autoform-googleplace.js'
  ], 'client');
  api.addAssets([
    'powered_by_google_on_white.png'
  ],'client');
});

Package.onTest(function(api) {
  api.use('tinytest');
  api.use('lukemadera:autoform-googleplace');
  api.addFiles('lukemadera_autoform-googleplace-tests.js');
});


lukemadera:autoform-googleplace.js
var VAL ={};
var OPTS ={};

var CLASSES ={
  predictions: 'hidden'
};

var ELES ={
  input: false,
  googleAttribution: false
};

Session.set('afGooglePlacePredictions', []);
Session.set('afGooglePlaceClasses', CLASSES);

var afGooglePlace ={

  updatePlace: function(val, place, params) {
    var self =this;
    var loc ={
      lat: '',
      lng: '',
      fullAddress: val,
      street: '',
      city: '',
      state: '',
      zip: '',
      country: ''
    };
    
    if(place && place.geometry && place.geometry.location) {
      loc =self.parseGoogleAddressComponent(place.address_components, {});

      loc.lat =place.geometry.location.lat();
      loc.lng =place.geometry.location.lng();
      loc.fullAddress =(place.formatted_address) ? place.formatted_address : "";
    }

    // ele.googleplace('val', loc);
    // this.value =loc;
    VAL =loc;
    return VAL;
  },

  /**
  Google returns lots of results and "address types" but the first one is usually the most specific so use that one to parse out the city and state.
  https://developers.google.com/maps/documentation/geocoding/?csw=1#ReverseGeocoding
  https://developers.google.com/maps/documentation/geocoding/?csw=1#Types
  For the United States, here's the mapping:
  - [street_number] + [route] --> street
  - locality --> city
  - administrative_area_level_1 --> state
  - postal_code --> zip code
  - country --> country
  @toc 2.
  @method parseGoogleAddressComponent
  @param {Object} addressComponents The raw result from google maps
  @return {Object} address The formatted address
  */
  parseGoogleAddressComponent: function(addressComponents, params) {
    var address ={};
    //go through all address components and pull out the matching types and map them to what we want (city, state)
    var map ={
      'street_number': 'street',
      'route': 'street',
      'locality': 'city',
      'administrative_area_level_1': 'state',
      'postal_code': 'zip',
      'country': 'country'
    };
    var ii, xx;
    for(ii =0; ii<addressComponents.length; ii++) {
      for(xx in map) {
        //if have a map type we want
        if(addressComponents[ii].types.indexOf(xx) >-1) {
          //have to join street number and route together
          if((xx ==='street_number' || xx ==='route') && address.street !==undefined) {
            //prepend
            if(xx ==='street_number') {
              address.street =addressComponents[ii].short_name + ' ' +address.street;
            }
            //append
            else if(xx ==='route') {
              address.street =address.street + ' ' + addressComponents[ii].short_name;
            }
          }
          else {
            if(xx ==='locality') {
              address[map[xx]] =addressComponents[ii].long_name;
            }
            else {
              address[map[xx]] =addressComponents[ii].short_name;
            }
          }
        }
      }
    }

    return address;
  },

  getPlace: function(inputVal, placePrediction, params) {
    var self =this;
    var placeService = new google.maps.places.PlacesService(ELES.googleAttribution);
    placeService.getDetails({placeId: placePrediction.place_id}, function(place, status) {
      if(status == google.maps.places.PlacesServiceStatus.OK) {
        var val =self.updatePlace(placePrediction.description, place, {});
        ELES.input.value =val.fullAddress;
        self.hide();
      }
      else {
        alert('google maps PlacesService error getting place with placeId: '+placeId);
      }
    });
  },

  updatePredictions: function(predictions, params) {
    Session.set('afGooglePlacePredictions', predictions);
    this.show({});
  },

  hide: function(params) {
    CLASSES.predictions ='hidden';
    Session.set('afGooglePlaceClasses', CLASSES);
  },

  show: function(params) {
    CLASSES.predictions ='visible';
    Session.set('afGooglePlaceClasses', CLASSES);
  },

  // getDropdownCoords: function(ele, params) {
  //   var rect = ele.getBoundingClientRect();
  //   // these are relative to the viewport
  //   var top = rect.top;
  //   var left = rect.left;
  //   return {
  //     top: top,
  //     left: left,
  //     bottom: top + ele.offsetHeight,
  //     right: left + ele.offsetWidth
  //   };
  // }

};

AutoForm.addInputType("googleplace", {
  template: "afGooglePlace",
  valueIn: function(val) {
    if(typeof(val) ==='string') {
      val ={
        fullAddress: val
      }
    }
    VAL =val;
    return val;
  },
  valueOut: function() {
    // return this.googleplace('val');
    // return this.value;
    return VAL;
  }
});

Template.afGooglePlace.rendered =function() {
  OPTS =EJSON.clone(this.data.atts.opts);
  if(OPTS ===undefined) {
    OPTS ={
      type: 'service'
    };
  }

  // var self =this;
  var ele =this.find('input');
  ELES.input =ele;
  ELES.googleAttribution =this.find('div.lm-autoform-google-place-attribution');
  var eleDropdown =this.find('div.lm-autoform-google-place-predictions');
  var types =[];    //either [blank] or one or more of: 'establishment', 'geocode'
  var componentRestrictions ={country:'us'};
  var options = {
    //bounds: defaultBounds,
    types: types,
    componentRestrictions: componentRestrictions
  };

  if(OPTS.type =='standard') {
    //standard autocomplete
    var autocomplete = new google.maps.places.Autocomplete(ele, options);

    google.maps.event.addListener(autocomplete, 'place_changed', function() {
      afGooglePlace.updatePlace(ele.value, autocomplete.getPlace(), {});
    });
  }
  else {
    //autocompleteservice (custom)
    // http://stackoverflow.com/questions/14414445/google-maps-api-v3-cant-geocode-autocompleteservice-predictions
    // https://developers.google.com/maps/documentation/javascript/examples/place-details
    // http://stackoverflow.com/questions/14343965/google-places-library-without-map
    var autocompleteService =new google.maps.places.AutocompleteService();
    ele.onkeyup =function(evt, params) {
      options.input =ele.value;
      if(!options.input.length) {
        afGooglePlace.hide({});
      }
      else {
        autocompleteService.getPlacePredictions(options, function(predictions, status) {
          if(status != google.maps.places.PlacesServiceStatus.OK) {
            // alert(status);
            afGooglePlace.hide({});
            return;
          }
          else {
            afGooglePlace.updatePredictions(predictions, {});
          }
        });
      }
    };

    // ele.onblur =function(evt, params) {
    //   var coords =afGooglePlace.getDropdownCoords(eleDropdown, {});
    //   console.log(coords);    //TESTING
    //   console.log(evt);
    //   afGooglePlace.hide({});
    // };

    ele.onfocus =function(evt, params) {
      if(ele.value.length) {
        afGooglePlace.show({});
      }
    };

  }
};

Template.afGooglePlace.helpers({
  //fix to avoid error for passed in object
  // - https://github.com/aldeed/meteor-autoform-bs-datepicker/issues/3
  // - https://github.com/aldeed/meteor-autoform-bs-datepicker/commit/3977aa69b61152cf8c0f731a11676b087d2ec9df
  atts: function() {
    var atts =EJSON.clone(this.atts);
    delete atts.opts;
    return atts;
  },
  classes: function() {
    return Session.get('afGooglePlaceClasses');
  }
});

Template.afGooglePlacePredictions.helpers({
  predictions: function() {
    return Session.get('afGooglePlacePredictions');
  }
});

Template.afGooglePlacePredictions.events({
  'click .lm-autoform-google-place-prediction-item': function(evt, template) {
    afGooglePlace.getPlace(ELES.input.value, this, {});
  }
});

lukemadera_autoform-googleplace.js
/**

@param {Object} [atts.opts]
  @param {String} [type ='service'] Set to 'googleUI' to use the Google Place UI (will NOT work on all mobile devices, especially iOS with 3rd party keyboards)
  @param {Boolean} [stopTimeoutOnKeyup =false] Set to false to keep running the timeout that auto-triggers showing predictions for 3rd party iOS keyboards where the keyup event does not fire. For performance, this defaults to true to kill the timeout as soon as we get a keyup event. This usually works, however, if the user switches back and forth between a 3rd party keyboard and a regular one, this will cause the 3rd party keyboard to NOT work anymore since a keyup event was registered (on the regular keyboard). So if you want this to be foolproof - albeit worse for performance, set this to false.
  @param {Object} [googleOptions] The google autocomplete options, such as:
    @param {Array} [types =[]]
    @param {Object} [componentRestrictions ={}]
*/

var VAL ={};    //one per instid

var afGooglePlace ={

  updatePlace: function(templateInst, val, place, params) {
    var self =this;
    var loc ={
      lat: '',
      lng: '',
      geometry: {
        type: "Point",
        coordinates: [ 0,0 ]
      },
      fullAddress: val,
      street: '',
      city: '',
      state: '',
      zip: '',
      country: '',
      placeId: ''
    };
    
    if(place && place.geometry && place.geometry.location) {
      loc =self.parseGoogleAddressComponent(place.address_components, {});

      loc.lat =place.geometry.location.lat();
      loc.lng =place.geometry.location.lng();
      loc.geometry = {
        type: "Point",
        coordinates: [ place.geometry.location.lng(), place.geometry.location.lat() ]
      };
      loc.fullAddress =(place.formatted_address) ? place.formatted_address : "";
      loc.placeId =place.place_id;
    }

    // ele.googleplace('val', loc);
    // this.value =loc;
    var instid =templateInst.data.atts['data-schema-key'];
    VAL[instid] =loc;
    return VAL[instid];
  },

  /**
  Google returns lots of results and "address types" but the first one is usually the most specific so use that one to parse out the city and state.
  https://developers.google.com/maps/documentation/geocoding/?csw=1#ReverseGeocoding
  https://developers.google.com/maps/documentation/geocoding/?csw=1#Types
  For the United States, here's the mapping:
  - [street_number] + [route] --> street
  - locality --> city
  - administrative_area_level_1 --> state
  - postal_code --> zip code
  - country --> country
  @toc 2.
  @method parseGoogleAddressComponent
  @param {Object} addressComponents The raw result from google maps
  @return {Object} address The formatted address
  */
  parseGoogleAddressComponent: function(addressComponents, params) {
    var address ={};
    //go through all address components and pull out the matching types and map them to what we want (city, state)
    var map ={
      'street_number': 'street',
      'route': 'street',
      'locality': 'city',
      'administrative_area_level_1': 'state',
      'postal_code': 'zip',
      'country': 'country'
    };
    var ii, xx;
    for(ii =0; ii<addressComponents.length; ii++) {
      for(xx in map) {
        //if have a map type we want
        if(addressComponents[ii].types.indexOf(xx) >-1) {
          //have to join street number and route together
          if((xx ==='street_number' || xx ==='route') && address.street !==undefined) {
            //prepend
            if(xx ==='street_number') {
              address.street =addressComponents[ii].short_name + ' ' +address.street;
            }
            //append
            else if(xx ==='route') {
              address.street =address.street + ' ' + addressComponents[ii].short_name;
            }
          }
          else {
            if(xx ==='locality') {
              address[map[xx]] =addressComponents[ii].long_name;
            }
            else {
              address[map[xx]] =addressComponents[ii].short_name;
            }
          }
        }
      }
    }

    return address;
  },

  getPlace: function(templateInst, placePrediction, params) {
    var self =this;
    var placeService = new google.maps.places.PlacesService(templateInst.eles.googleAttribution);
    placeService.getDetails({placeId: placePrediction.place_id}, function(place, status) {
      if(status == google.maps.places.PlacesServiceStatus.OK) {
        var val =self.updatePlace(templateInst, placePrediction.description, place, {});
        templateInst.eles.input.value =val.fullAddress;
        templateInst.timeouthack.lastVal =templateInst.eles.input.value;   //update for next time
        self.hide(templateInst, {});
      }
      else {
        alert('google maps PlacesService error getting place with placeId: '+placeId);
      }
    });
  },

  /**
  Selects either the given prediction OR, if prediction is null or undefined
   tries to select the currently selected (keyboard) prediction
  */
  choosePrediction: function(templateInst, prediction) {
    templateInst.timeouthack.trigJustSelectedVal =true;
    setTimeout(function() {
      templateInst.timeouthack.trigJustSelectedVal =false;
    }, templateInst.timeouthack.times.selectedVal);

    if(!prediction) {
      var predictionsSelected =templateInst.predictionsSelected.get();
      var predictions =templateInst.predictions.get();
      prediction =predictions[predictionsSelected.index];
    }
    if(prediction) {
      afGooglePlace.getPlace(templateInst, prediction, {});
    }
  },

  /**
  @param {Object} options
    @param {String} input Element value
  @param {Object} params
    @param {Boolean} [noShow] True to NOT display predictions
    @param {Boolean} [setVal] True to also set the value (i.e. for init)
  */
  getPredictions: function(templateInst, options, params) {
    var self =this;
    templateInst.autocompleteservice.getPlacePredictions(options, function(predictions, status) {
      if(status != google.maps.places.PlacesServiceStatus.OK) {
        // alert(status);
        self.hide(templateInst, {});
        return;
      }
      else {
        self.updatePredictions(templateInst, predictions, params);
        if(params.setVal) {
          self.getPlace(templateInst, predictions[0], {});
        }
      }
    });
  },

  /**
  @param {Object} params
    @param {Boolean} [noShow] True to NOT show dropdown
  */
  updatePredictions: function(templateInst, predictions, params) {
    //add display key for tracking selected prediction for keyboard support
    predictions.map(function(prediction) {
      prediction.xDisplay ={
        selected: ''
      };
      return prediction;
    });
    //reset index
    var predictionsSelected =templateInst.predictionsSelected.get();
    predictionsSelected.index =-1;
    templateInst.predictionsSelected.set(predictionsSelected);

    templateInst.predictions.set(predictions);
    if(params.noShow ===undefined || !params.noShow) {
      this.show(templateInst, {});
    }
  },

  /**
  Handles (keyboard) up / down to change currently selected prediction in dropdown
  @param {String} change One of 'prev', 'next'
  */
  updatePredictionsSelected: function(templateInst, change, params) {
    var predictionsSelected =templateInst.predictionsSelected.get();
    var predictions =templateInst.predictions.get();
    var changed =false;
    if(change ==='prev' && predictionsSelected.index >0) {
      predictionsSelected.index--;
      changed =true;
    }
    else if(change ==='next' && predictionsSelected.index <(predictions.length-1)) {
      predictionsSelected.index++;
      changed =true;
    }
    if(changed) {
      templateInst.predictionsSelected.set(predictionsSelected);
      predictions.map(function(prediction, index) {
        if(index ===predictionsSelected.index) {
          prediction.xDisplay.selected ='selected';
        }
        else {
          prediction.xDisplay.selected ='';
        }
      });
      templateInst.predictions.set(predictions);
    }
  },

  hide: function(templateInst, params) {
    var classes =templateInst.classes.get();
    classes.predictions ='hidden';
    templateInst.classes.set(classes);
  },

  show: function(templateInst, params) {
    var classes =templateInst.classes.get();
    classes.predictions ='visible';
    templateInst.classes.set(classes);
  },

  // getDropdownCoords: function(ele, params) {
  //   var rect = ele.getBoundingClientRect();
  //   // these are relative to the viewport
  //   var top = rect.top;
  //   var left = rect.left;
  //   return {
  //     top: top,
  //     left: left,
  //     bottom: top + ele.offsetHeight,
  //     right: left + ele.offsetWidth
  //   };
  // }

};

AutoForm.addInputType("googleplace", {
  template: "afGooglePlace",
  valueIn: function(val) {
    //will convert to display value later after set / extend opts
    return val;
  },
  valueOut: function() {
    var instid =this.attr('data-schema-key');
    return VAL[instid];
  }
});

Template.afGooglePlace.created =function() {
  this.autocompleteservice =new google.maps.places.AutocompleteService();
  this.timeouthack ={
    lastVal: '',
    trigJustSelectedVal: false,
    times: {
      keyup: 500,
      selectedVal: 1000   //must be LONGER than keyup timeout time
    },
    trigs: {
      keyup: false
    }
  };
  this.eles ={
    input: false,
    googleAttribution: false
  };
  this.opts ={};

  this.predictions =new ReactiveVar([]);
  this.predictionsSelected =new ReactiveVar({
    index: -1
  });
  this.classes =new ReactiveVar({
    predictions: 'hidden'
  });
};

Template.afGooglePlace.rendered =function() {
  var templateInst =this;
  //reset (any past instances)
  if(templateInst.timeouthack.trigs.keyup) {
    clearTimeout(templateInst.timeouthack.trigs.keyup);
  }

  var optsDefault ={
    type: 'service',
    // stopTimeoutOnKeyup: true
    stopTimeoutOnKeyup: false
  };
  var xx;
  templateInst.opts =EJSON.clone(this.data.atts.opts);
  if(templateInst.opts ===undefined) {
    templateInst.opts =EJSON.clone(optsDefault);
  }
  else {
    //extend
    for(xx in optsDefault) {
      if(templateInst.opts[xx] ===undefined) {
        templateInst.opts[xx] =optsDefault[xx];
      }
    }
  }

  // var self =this;
  var ele =this.find('input');

  //temporarily set value (just need it to be an object; will set properly later)
  var val =ele.value;
  if(typeof(val) ==='string') {
    val ={
      fullAddress: val
    };
  }
  var instid =templateInst.data.atts['data-schema-key'];
  VAL[instid] =val;

  templateInst.eles.input =ele;
  templateInst.eles.googleAttribution =this.find('div.lm-autoform-google-place-attribution');
  var eleDropdown =this.find('div.lm-autoform-google-place-predictions');
  var types =[];    //either [blank] or one or more of: 'establishment', 'geocode'
  var componentRestrictions ={};
  var options = {
    //bounds: defaultBounds,
    types: types,
    componentRestrictions: componentRestrictions
  };
  if(templateInst.opts.googleOptions !==undefined) {
    //extend
    for(xx in templateInst.opts.googleOptions) {
      options[xx] =templateInst.opts.googleOptions[xx];
    }
  }

  if(templateInst.opts.type =='googleUI') {
    //standard autocomplete
    var autocomplete = new google.maps.places.Autocomplete(templateInst.eles.input, options);

    google.maps.event.addListener(autocomplete, 'place_changed', function() {
      afGooglePlace.updatePlace(templateInst, templateInst.eles.input.value, autocomplete.getPlace(), {});
    });
  }
  else {
    //autocompleteservice (custom)
    // http://stackoverflow.com/questions/14414445/google-maps-api-v3-cant-geocode-autocompleteservice-predictions
    // https://developers.google.com/maps/documentation/javascript/examples/place-details
    // http://stackoverflow.com/questions/14343965/google-places-library-without-map

    //3rd party iOS mobile keyboards do NOT fire the events properly so we need to do a timeout. To handle this, we'll auto run the event handlers with a periodic timeout but then if we DO get an event, we'll stop the timeout since we have now "detected" the keyboard is not 3rd party (or at least that it's working)
    var eventsFiring =false;    //default - once we get an event, we'll reset this
    templateInst.timeouthack.lastVal =templateInst.eles.input.value;   //save to compare to track if a change happened

    var timeoutTriggerEvents =function(params) {
      templateInst.timeouthack.trigs.keyup =setTimeout(function() {
        if(!eventsFiring) {
          if(!templateInst.timeouthack.trigJustSelectedVal) {
            handleKeyup(false, {});
          }
          timeoutTriggerEvents({});   //call again
        }
      }, templateInst.timeouthack.times.keyup);
    };

    var handleKeyup =function(evt, params) {
      // options.input =ele.value;    //does not work - init / instance id issue
      options.input =templateInst.eles.input.value;
      if(!options.input.length) {
        afGooglePlace.hide(templateInst, {});
      }
      else {
        if(templateInst.eles.input.value !==templateInst.timeouthack.lastVal) {
          afGooglePlace.getPredictions(templateInst, options, {});
          templateInst.timeouthack.lastVal =templateInst.eles.input.value;   //update for next time
        }
      }
    };

    var handleFocus =function(evt, params) {
      if(templateInst.eles.input.value.length) {
        afGooglePlace.show(templateInst, {});
      }
    };

    templateInst.eles.input.onkeydown =function(evt, params) {
      if(evt.keyCode ===13) {
        return false;
      }
    };

    templateInst.eles.input.onkeyup =function(evt, params) {
      //up arrow
      if(evt.keyCode ===38) {
        afGooglePlace.show(templateInst, {});
        afGooglePlace.updatePredictionsSelected(templateInst, 'prev', {});
      }
      //down arrow
      else if(evt.keyCode ===40) {
        afGooglePlace.show(templateInst, {});
        afGooglePlace.updatePredictionsSelected(templateInst, 'next', {});
      }
      //enter
      else if(evt.keyCode ===13) {
        afGooglePlace.choosePrediction(templateInst, null);
      }
      //escape
      else if(evt.keyCode ===27) {
        afGooglePlace.hide(templateInst, {});
      }
      else {
        if(templateInst.opts.stopTimeoutOnKeyup) {
          eventsFiring =true;   //update trigger; stop timeout
        }
        handleKeyup(evt, params);
      }
    };

    // templateInst.eles.input.onblur =function(evt, params) {
    //   var coords =afGooglePlace.getDropdownCoords(eleDropdown, {});
    //   console.log(coords);    //TESTING
    //   console.log(evt);
    //   afGooglePlace.hide(templateInst, {});
    // };

    templateInst.eles.input.onfocus =function(evt, params) {
      handleFocus(evt, params);
    };

    //start timeout going
    timeoutTriggerEvents({});

    //init
    if(templateInst.eles.input.value && templateInst.eles.input.value.length) {
      options.input =templateInst.eles.input.value;
      afGooglePlace.getPredictions(templateInst, options, {noShow:true, setVal:true});
    }
    else {
      afGooglePlace.hide(templateInst,{});  
    }

  }
};

Template.afGooglePlace.helpers({
  //fix to avoid error for passed in object
  // - https://github.com/aldeed/meteor-autoform-bs-datepicker/issues/3
  // - https://github.com/aldeed/meteor-autoform-bs-datepicker/commit/3977aa69b61152cf8c0f731a11676b087d2ec9df
  atts: function() {
    var atts =EJSON.clone(this.atts);
    delete atts.opts;
    return atts;
  },
  classes: function() {
    return Template.instance().classes.get();
  }
});

Template.afGooglePlace.helpers({
  predictions: function() {
    return Template.instance().predictions.get();
  }
});

Template.afGooglePlace.events({
  'click .lm-autoform-google-place-prediction-item': function(evt, template) {
    afGooglePlace.choosePrediction(template, this);
  }
});